modules/packages/Collection.chpl

```diff
@@ -18,118 +18,212 @@
 
-      var capturedRetval = c.remove()
+      ::
 
-    **BUG:** Loop Invariant Code Motion causes undefined behavior if assigned to a
-    variable declared outside of loop. Issue: #7003
+        var capturedRetval = c.remove()
 
-    **FIX:** Use the `--no-loop-invariant-code-motion` to disable LICM.
-    Otherwise, just make sure you always capture the return value inside of a loop
-    in a variable not declared outside of loop...
+      **BUG:** Loop Invariant Code Motion causes undefined behavior if assigned to a
+      variable declared outside of loop. Issue: #7003
 
-    ::
+      **FIX:** Use the `--no-loop-invariant-code-motion` to disable LICM.
+      Otherwise, just make sure you always capture the return value inside of a loop
+      in a variable not declared outside of loop...
 
-      for i in 1 .. N {
-        var retval = c.remove();
-      }
+      .. code-block:: chapel
 
-  */
-  inline proc remove() : (bool, eltType) {
-    halt("'proc remove() : (bool, eltType)' is not supported...");
-  }
+        for i in 1 .. N {
+          var retval = c.remove();
+        }
+      
+      **BUG:** Sometimes the compiler will produce an internal error related to 'visibility blocks',
+      this can be resolved by declaring the return type rather than letting the compiler infer it.
+      The exact cause has yet to be found.
 
-  /*
-    Determine whether an element exists in this collection.
-  */
-  inline proc contains(elt : eltType) : bool {
-    halt("'proc contains(elt : eltType) : bool' is not supported...");
-  }
+      .. code-block:: chapel
+  
+        var retval : (bool, eltType) = c.remove();
 
-  /*
-    Clears all elements in this collection.
-  */
-  inline proc clear() {
-    while !remove()[1] do ;
-  }
+    */
+    proc remove() : (bool, eltType) {
+      halt("'proc remove() : (bool, eltType)' is not supported...");
+    }
 
-  /*
-    Check if this data structure is empty.
-  */
-  inline proc isEmpty() : bool {
-    return getSize() == 0;
-  }
+    /*
+      Removes `nElts` elements from this data structure, returning them as an array.
+      If the data structure fails to produce a new element, we cease and shrink the array
+      to its appropriate size and return that. 
+    */
+    proc removeBulk(nElts : int) {
+      var dom = {0..#nElts};
+      var arr : [dom] eltType;
+      var idx = 0;
 
-  /*
-    Syntactic sugar for `getSize`.
-  */
-  inline proc length : int {
-    return getSize();
-  }
+      for 1 .. nElts {
+        var (hasElt, elt) = remove();
+        if !hasElt {
+          dom = {0..#idx};
+          break;
+        }
 
-  /*
-    Syntactic sugar for `getSize`.
-  */
-  inline proc size : int {
-    return getSize();
-  }
+        arr[idx] = elt;
+        idx += 1;
+      }
 
-  /*
-    Obtain the number of elements contained in this collection.
-  */
-  inline proc getSize() : int {
-    halt("'proc size() : int' is not supported...");
+      return arr;

```

Comment:
I wonder if you can return an array slice safely here. It might deliver better performance as you probably won't need to shrink the array, which is obviously costly.

Reply:
Github isn't showing me the related code in a way that I understand, but generally speaking array slices return by value right now.